### PR TITLE
Implement is_array_like

### LINKED ISSUE #12764: CLN: internal array detection
so need to add some functions to replace lots of instance checking for compat in `core/common.py`

```
def is_array_like(arr):
    return isinstance(arr, np.ndarray)
```

analagously to `is_list_like` functions.

These will isolate code changes later.

### PR DESCRIPTION
**Pull request by sornars**
- [ ] closes #12764 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

Is it worth using this PR to consolidate how the common package is imported across modules? Some files use `from pandas.core.common import *` style whereas others use `import pandas.core.common as com`. Some even use both styles in the same file - is there a style preference and should I go ahead and fix this now?

### CODE REVIEW

**Comment by jreback** [comment]
you will have to rebase after I finish reorging types. https://github.com/jreback/pandas/commit/3ef6e8e0ef52476c4b6769263d18ebc7fbf8de6f

should be done in next day or 2.

**Comment by shoyer** [comment]
I'm not sure this is a good idea -- let's finish the discussion of what this means back in #12764 before doing further work here. In particular, most of the code protected by these checks is still specific to NumPy arrays, so using `is_array_like` instead of `isinstance` obfuscates what is going on.

**Comment by jreback** [comment]
@shoyer you are diving too deep here. This is just an alias at this moment. Later it will need to be changed to actually do something other than ``isinsanstance(arr, np.ndarray)``

**Comment by shoyer** [comment]
Just to be clear, once we change `is_array_like`, we'll need to audit every place where it's called. What value do we get from changing it preemptively?

**Comment by jreback** [comment]
can you rebase / update?

though a lot has changed. ok with closing too.

**Comment by sornars** [comment]
Will update and push something up; hopefully by the end of this weekend.

**Comment by jreback** [comment]
@wesm any objections?

File: pandas/types/common.py [hunk 1]
```diff
@@ -307,6 +307,10 @@ def _coerce_to_dtype(dtype):
     return dtype
 
 
+def is_arraylike(arg):
+    return isinstance(arg, np.ndarray)
```
Comment (jreback) [inline pandas/types/common.py]: can you add a doc-string

and some tests

**Comment by jreback** [comment]
@sornars also, can you add to pandas.api.types as well

**Comment by shoyer** [comment]
Per the discussion in #12764, I am -1 on calling this function `is_array_like`. We could call this `is_array`, but `isinstance(x, np.ndarray)` is quite short and readable already. Adding a layer of indirection would only make code using this check less obvious, so I am -0 even on `is_array`.

**Comment by jreback** [comment]
@shoyer so when we have ``pa.Array`` (the 'new' array object)

you want to go an just change ``isinstance(arr, (np.ndarray, pa.Array))``. you think that is more idiomatic / clearer? really?

**Comment by shoyer** [comment]
When we have `pa.Array`, then yes we will probably want to add `is_array` and use it in many cases. But there is no way that `pa.Array` is a 100% drop in replacement for `numpy.ndarray` (e.g., for`dtype` and N-dimensional functionality), so switching every single use of `isinstance(arr, np.ndarray)` to `isinstance(arr, (np.ndarray, pa.Array))` (via `is_array`) is assuredly not going to be what we want.

**Comment by wesm** [comment]
I'm OK with this function *internally*. Since I predict we'll see fewer and fewer raw ndarrays floating around pandas internals (instead they would be boxed in pandas.Array objects) I expect a lot of this code to go away in the future

**Comment by jorisvandenbossche** [comment]
I don't really like the `is_arraylike` name (apart from that it should be `is_array_like` to be consistent with other functions). "array-like" has for me a clear meaning of something that is "asarray"-able to an array, so including eg lists. 
So if we include this function, I think `is_array` or `is_np_array` would be a better fit. However, I follow @shoyer that `isinstance(x, np.ndarray)` is very explicit and already readable as well.

File: pandas/api/tests/test_api.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ class TestTypes(Base, tm.TestCase):
                'is_dict_like', 'is_iterator',
                'is_list_like', 'is_hashable',
                'is_named_tuple', 'is_sequence',
-               'pandas_dtype']
+               'pandas_dtype', 'is_arraylike']
```
Comment (jreback) [inline pandas/api/tests/test_api.py]: so remove from here

**Comment by jreback** [comment]
@sornars 
change to ``is_array_like`` (and I think the meaning in pandas is quite clear; this is very distinct from ``is_list_like``). 

pls document that this is an internal function only (IOW, not exposing in the api).

some flake8 errors as well.

**Comment by wesm** [comment]
We could change the function name to `is_ndarray` to be more clear. My understanding is that the purpose of this is to reduce code duplication / smell, not to expand the definition of what is like an array? For example, Series are array-like, but we wouldn't want to put them in this function?

**Comment by jreback** [comment]
well, I think this will make things easier by having the ability to change the array definition in one place. This is meant specifically for testing ndarray (and not list and not Series) things, e.g. the definition *mostly* could eventually be:

```
def is_array(arr):
    return isinstance(arr, (np.ndarray, pa.Array))
```

with specilizations like:

```
def is_ndarray(arr):
    return isinstance(arr, np.ndarray)

def is_paarray(arr):
    return isnstance(arra, pa.Array)
```

this is really for consistency as we use ``is_*`` for everything else internally.

**Comment by wesm** [comment]
In the future that may make sense, but I think we would want to go through and decide specifically where we want to expand the definition of what kind of object passes the `if` condition (i.e. whether to change `is_ndarray` in a particular case to `is_array_zerocopyable`)

**Comment by jreback** [comment]
fair enough

thanks for the effort @sornars

**Pull request closed without merging**